sqlglot/helper.py

```diff
@@ -64,14 +92,26 @@ def apply_index_offset(expressions, offset):
         logger.warning("Applying array index offset (%s)", offset)
         expression.args["this"] = str(int(expression.args["this"]) + offset)
         return [expression]
+
     return expressions
 
 
-def camel_to_snake_case(name):
+def camel_to_snake_case(name: str) -> str:
+    """Converts `name` from camelCase to snake_case and returns the result."""
     return CAMEL_CASE_PATTERN.sub("_", name).upper()
 
 
-def while_changing(expression, func):
+def while_changing(expression: t.Optional[Expression], func: t.Callable) -> t.Optional[Expression]:
```

Comment:
I think this is what you want:

```python
E = t.TypeVar("E", bound=Expression)

def while_changing(expression: t.Optional[Expression], func: t.Callable[[Expression], E]) -> E:
```

The TypeVar might be overkill. But in any case, I think we should annotate the Callable.